components/api/docker_registry/registry.bal

```diff
@@ -0,0 +1,129 @@
+// limitations under the License
+//
+// ------------------------------------------------------------------------
+
+import ballerina/config;
+import ballerina/http;
+import ballerina/io;
+import ballerina/log;
+
+http:Client dockerRegistryClientEP = new(config:getAsString("docker.registry.url"), config = {
+    secureSocket: {
+        trustStore: {
+            path: config:getAsString("security.truststore"),
+            password: config:getAsString("security.truststorepass")
+        },
+        verifyHostname: false
+    }
+});
+
+public function getResponseFromManifestAPI(string orgName, string imageName, string artifactVersion = "", string digest = "", string bearerToken = "")
+returns http:Response? | error?  {
+    log:printDebug(io:sprintf("Invoking docker registry API. orgName: %s, imageName: %s, artifactVersion: %s, digest: %s",
+    orgName, imageName, artifactVersion, digest));
+
+    http:Request dockerRegistryRequest = new;
+    dockerRegistryRequest.addHeader("Authorization", "Bearer " + bearerToken);
+
+    if (artifactVersion != "") {
+        string getManifestEndPoint = io:sprintf("/v2/%s/%s/manifests/%s", orgName, imageName, artifactVersion);
+        log:printDebug("Calling getDockerManifestDigest end point");
+        return dockerRegistryClientEP->get(getManifestEndPoint, message = dockerRegistryRequest);
+    } else if (digest != "") {
+        string deleteEndPoint = io:sprintf("/v2/%s/%s/manifests/%s", orgName, imageName, digest);
+        log:printDebug("Calling deleteDockerManifest end point");
+        return dockerRegistryClientEP->delete(deleteEndPoint, dockerRegistryRequest);
+    }
+}
+
+public function getManifestDigest(string orgName, string imageName, string artifactVersion, string bearerToken = "", string userId, string token)
+returns string | error {
+    var responseForGetManifest = getResponseFromManifestAPI(orgName, imageName, artifactVersion = artifactVersion, bearerToken = bearerToken);
+    if (responseForGetManifest is http:Response) {
+        log:printDebug(io:sprintf("Received status code for getManifestDigest request: %d", responseForGetManifest.statusCode));
+        if (responseForGetManifest.statusCode == http:UNAUTHORIZED_401) {
+            var payload = responseForGetManifest.getJsonPayload();
+            if (payload is json) {
+                log:printDebug(io:sprintf("Received payload from docker registry for getManifestDigest request: %s", payload));
+                string registryScopeForGetManifest = buildRegistryScope(payload);
+                log:printDebug(io:sprintf("Registry scope for getManifest : %s", registryScopeForGetManifest));
+
+                string tokenToGetManifestDigest = getTokenFromDockerAuth(userId, token, registryScopeForGetManifest);
+                log:printDebug("Retrived a token to get manifest digest");
+                return getManifestDigest(orgName, imageName, artifactVersion, bearerToken = tokenToGetManifestDigest, userId, token);
```

Comment:
This might cause an infinite recursion.

Eg:- The API has cached tokens and if the token is revoked by the user, this might go to a infinite recursion.